src/widgets/taskbarnotifier.cpp

```diff
@@ -0,0 +1,251 @@
+#ifdef USE_DBUS
+    ,
+    isServiceAvailable_(false), desktopPath_(QLatin1String())
+{
+    isServiceAvailable_ = checkDBusSeviceAvailable();
+}
+#elif defined(Q_OS_WINDOWS)
+{
+}
+#endif
+
+#ifdef USE_DBUS
+bool TaskBarNotifier::Private::checkDBusSeviceAvailable()
+{
+    const auto services = QDBusConnection::sessionBus().interface()->registeredServiceNames().value();
+    for (const auto &service : services) {
+        if (service.contains(ULAUNCHER_SERV, Qt::CaseInsensitive))
+            return true;
+    }
+    return false;
+}
+
+void TaskBarNotifier::Private::sendDBusSignal(bool isVisible, uint number)
+{
+    if (isServiceAvailable_) {
+        QDBusMessage signal = QDBusMessage::createSignal(ULAUNCHER_PATH, ULAUNCHER_IFACE, ULAUNCHER_CMD);
+        signal << desktopPath_;
+        QVariantMap args;
+        args["count-visible"] = isVisible;
+        args["count"]         = number;
+        args["urgent"]        = urgent_;
+        signal << args;
+        QDBusConnection::sessionBus().send(signal);
+    }
+}
+
+void TaskBarNotifier::Private::setDesktopPath(const QString &appName)
+{
+#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
+    desktopPath_ = QLatin1String("application://%1").arg(appName);
+#else
+    desktopPath_ = QLatin1String("application://%1.desktop").arg(appName);
+#endif
+}
+
+#elif defined(Q_OS_WINDOWS)
+QImage TaskBarNotifier::Private::makeIconCaption(const QImage &image, const QString &number) const
+{
+    if (!image.isNull()) {
+        auto   imSize    = image.size() * devicePixelRatio_;
+        QImage img       = image;
+        auto   letters   = number.length();
+        auto   text      = QStaticText((letters < 3) ? number : "99+");
+        auto   textDelta = (letters <= 2) ? 3 : 4;
+
+        QPainter p(&img);
+        p.setRenderHints(QPainter::Antialiasing | QPainter::TextAntialiasing);
+        auto font   = QFont("Times", imSize.height() / textDelta, QFont::Bold);
+        auto fm     = QFontMetrics(font);
+        auto fh     = fm.height();
+        auto fw     = fm.horizontalAdvance(text.text());
+        auto radius = fh / 2;
+
+        Qt::BrushStyle style = Qt::SolidPattern;
+        QBrush         brush(Qt::red, style);
+        p.setBrush(brush);
+        p.setPen(QPen(Qt::NoPen));
+        QRect rect(imSize.width() - fw - radius, 0, fw + radius, fh);
+        p.drawRoundedRect(rect, radius, radius);
+
+        p.setFont(font);
+        p.setPen(QPen(Qt::white));
+        auto offset = rect.width() / ((letters + 1) * 2);
+        p.drawStaticText(rect.x() + offset, rect.y(), text);
+
+        p.end();
+        return img;
+    }
+    return QImage();
+}
+
+void TaskBarNotifier::Private::setParentIcon(const QImage &image, uint count)
+{
+    if (image.isNull())
+        return;
+
+    QImage img;
+    if (count > 0 && urgent_) {
+        img = makeIconCaption(image, QString::number(count));
+    }
+    HICON icon = (img.isNull()) ? getHICONfromQImage(image) : getHICONfromQImage(img);
```

Comment:
I remember original code called DestroyHicon or something like this in the end.
memory leak?

Reply:
Fixed